src/codegen.c

```diff
@@ -0,0 +1,1238 @@
+#include "softfloat.h"
+
+#define SET_SIZE 1024
+
+typedef struct {
+    uint32_t table[SET_SIZE][32];
+} set_t;
+
+static inline uint32_t hash(uint32_t key)
+{
+    return (key >> 1) & (SET_SIZE - 1);
+}
+
+void set_reset(set_t *set)
+{
+    memset(set, 0, sizeof(set_t));
+}
+
+bool set_add(set_t *set, uint32_t key)
+{
+    uint32_t index = hash(key);
+    uint8_t count = 0;
+    while (set->table[index][count] != 0) {
+        if (set->table[index][count++] == key)
+            return false;
+    }
+
+    set->table[index][count] = key;
+    return true;
+}
+
+bool set_has(set_t *set, uint32_t key)
+{
+    uint32_t index = hash(key);
+    uint8_t count = 0;
+    while (set->table[index][count] != 0) {
+        if (set->table[index][count++] == key)
+            return true;
+    }
+    return false;
+}
+
+static bool insn_is_branch(uint8_t opcode)
+{
+    switch (opcode) {
+#define _(inst, can_branch) IIF(can_branch)(case rv_insn_##inst:, )
+        RISCV_INSN_LIST
+#undef _
+        return true;
+    }
+    return false;
+}
+typedef void (*gen_func_t)(riscv_t *, const rv_insn_t *, char *, uint32_t);
+static gen_func_t dispatch_table[rv_insn_qty];
+static char funcbuf[128] = {0};
+#define UPDATE_PC(inc)                          \
+    sprintf(funcbuf, "  rv->PC += %d;\n", inc); \
+    strcat(gencode, funcbuf);
+#define NEXT_INSN(target)                          \
+    sprintf(funcbuf, "  goto insn_%x;\n", target); \
+    strcat(gencode, funcbuf);
+#define RVOP(inst, code)                                            \
+    static void gen_##inst(riscv_t *rv UNUSED, const rv_insn_t *ir, \
+                           char *gencode, uint32_t pc)              \
+    {                                                               \
+        sprintf(funcbuf,                                            \
+                "insn_%x:\n"                                        \
+                "  rv->X[0] = 0;\n"                                 \
+                "  rv->csr_cycle++;\n",                             \
+                (pc));                                              \
+        strcat(gencode, funcbuf);                                   \
+        code;                                                       \
+        if (!insn_is_branch(ir->opcode)) {                          \
+            UPDATE_PC(ir->insn_len);                                \
+            NEXT_INSN(pc + ir->insn_len);                           \
+        }                                                           \
+    }
+
+RVOP(nop, {/* no operation */});
+
+RVOP(lui, {
+    sprintf(funcbuf, "  rv->X[%u] = %u;\n", ir->rd, ir->imm);
+    strcat(gencode, funcbuf);
+})
+
+RVOP(auipc, {
+    sprintf(funcbuf, "  rv->X[%u] = %u + rv->PC;\n", ir->rd, ir->imm);
+    strcat(gencode, funcbuf);
+})
+
+RVOP(jal, {
+    strcat(gencode, "  pc = rv->PC;\n");
+    UPDATE_PC(ir->imm);
+    if (ir->rd) {
+        sprintf(funcbuf, "  rv->X[%u] = pc + %u;\n", ir->rd, ir->insn_len);
```

Comment:
Don't invoke `sprintf` in each implementation of RISC-V instructions. Instead, wrap them in proper macros or inline functions.